### PR TITLE
dev/core#2634 [REF] Move membership date calc from v3 api to BAO

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
[REF] Move membership date calc from v3 api to BAO 

Before
----------------------------------------
code to calculated membership dates  for new membership is in the v3 api and in every form that creates memberships

After
----------------------------------------
calculated in the BAO for v3 api & membership batch entry form

Technical Details
----------------------------------------
This moves code to calculate membership dates for new memberships from the
v3 api Membership.create and the batch entry membership form to the membership bao.

I think more cleanup could follow this - in the api most obviously and
the BAO is a bit insane still

This has test cover in the v3 api membership test and in
CRM_Batch_Form_EntryTest

Comments
----------------------------------------
Note this was originally https://github.com/civicrm/civicrm-core/pull/20397 - it's been closed for 6 weeks waiting for me to devote the time to figure out the test failure & I'm going to discuss that here

Basically the failure was because in the BAO there was an assumption that is_override meant 'do not calculate missing dates OR the membership status' whereas the api understood it to mean ' do not calculate membership status'

In the end I settled on the api interpretation because

1) the api is the main way in which Membership.create is interacted with. There are a number of places in core that call the BAO directly but these generally pass in the dates already calculated and I've added pretty extensive coverage I believe.
2) all the evidence I can find is that 'is_override' refers to status, not dates

### CODE REVIEW

**Comment by colemanw** [comment]
This is good cleanup with solid test cover. Better in than out.

**Comment by eileenmcnaughton** [comment]
Yay thanks @colemanw

**Comment by KarinG** [comment]
May have resulted in https://lab.civicrm.org/dev/core/-/issues/2902 (Date fields are empty on memberships (created with pay later OR with an API call)).
